### DISCUSSION TITLE
Mermaid built in icons do not load

### DISCUSSION THREAD

**Discussion by alfredopalhares** [squidfunk/mkdocs-material#7723] [Q&A]
### Context

Hello,  first of all thank you so much for working on this project, it has provided me the best way to quickly deploy documentation that actually gets maintained and updated.

I am trying to use the [architecture-diagrams](https://mermaid.js.org/syntax/architecture.html), but whenever I try to use icons from iconify.design as described on the doc. 
```

### Bug description

Given a sample diagram like does not render icons:
```plain
architecture-beta
    group api(logos:aws-lambda)[API]

    service db(logos:aws-aurora)[Database] in api
    service disk1(logos:aws-glacier)[Storage] in api
    service disk2(logos:aws-s3)[Storage] in api
    service server(logos:aws-ec2)[Server] in api

    db:L -- R:server
    disk1:T -- B:server
    disk2:T -- B:db
```

![Screenshot From 2024-11-19 17-03-59](https://github.com/user-attachments/assets/ddf8bd0f-41f8-40f8-bf23-5904f9298b09)

In have tried to load the icons myself by adding this to `docs/src/js/register-mermaid-icons.js`:

```javascript
import mermaid from "mermaid";

mermaid.registerIconPacks([
  {
    name: "logos",
    loader: () =>
      fetch("https://unpkg.com/@iconify-json/logos/icons.json")
        .then((res) => {
          if (!res.ok) throw new Error("Failed to fetch icons");
          return res.json();
        })
        .then((icons) => {
          console.log("Icons loaded:", icons);
          return icons;
        })
        .catch((err) => console.error("Error loading icons:", err)),
  },
]);

// Initialize Mermaid
mermaid.initialize({
  startOnLoad: true,
  theme: "default", // or "dark", "neutral"
  securityLevel: "loose", // Needed for inline HTML in Mermaid
});

```

And load it from `mkdocs.yml`:
```yaml
extra_javascript:
  - path: src/js/register-mermaid-icons.js
    type: module
```

The codes does run, by mermad.registerIconPacks never runs.. what am I doing wrong ?

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)
-https://mermaid.js.org/syntax/architecture.html

### Reproduction

[9.5.44-mermaid-icons.zip](https://github.com/user-attachments/files/17819539/9.5.44-mermaid-icons.zip)

### Steps to reproduce

Build the reproduction sample or use the diagram and js provided on the bu

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material/), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/) using the [built-in info plugin](https://squidfunk.github.io/mkdocs-material/plugins/info/).

**Comment by facelessuser** [comment]
You may want to check out this thread: https://github.com/facelessuser/pymdown-extensions/discussions/2499

**Comment by alfredopalhares** [comment]
@facelessuser  I actually checked it, but I cannot reproduce that example for sure, I have checked your most up to date code too.

**Comment by facelessuser** [comment]
@alfredopalhares The main point of the conversation highlights two things. The first is that you need to register the icons with a name.

```js
     mermaid.registerIconPacks([
       {
         name: "simple-icons",
         loader: () =>
           fetch("https://unpkg.com/@iconify-json/logos/icons.json").then(res =>
             res.json()
           )
       }
     ])
```

And when you reference them, you need to use the same name you referenced them with:

```
architecture-beta
    group api(simple-icons:aws-lambda)[API]

    service db(simple-icons:aws-aurora)[Database] in api
    service disk1(simple-icons:aws-glacier)[Storage] in api
    service disk2(simple-icons:aws-s3)[Storage] in api
    service server(simple-icons:aws-ec2)[Server] in api

    db:L -- R:server
    disk1:T -- B:server
    disk2:T -- B:db
```

**Comment by squidfunk** [comment]
Converting to discussion - this is not officially supported.

**Comment by alfredopalhares** [comment]
Yes, I got that, but the part i am probably failing is the actually function, because on this example:

```javascript
import mermaid from "mermaid";

mermaid.registerIconPacks([
  {
    name: "logos",
    loader: () =>
      fetch("https://unpkg.com/@iconify-json/logos/icons.json")
        .then((res) => {
          if (!res.ok) throw new Error("Failed to fetch icons");
          return res.json();
        })
        .then((icons) => {
          console.log("Icons loaded:", icons);
          return icons;
        })
        .catch((err) => console.error("Error loading icons:", err)),
  },
]);

// Initialize Mermaid
mermaid.initialize({
  startOnLoad: true,
  theme: "default", // or "dark", "neutral"
  securityLevel: "loose", // Needed for inline HTML in Mermaid
});
```

The code inside the function never gets printed on the browser console. 
What am I doing wrong ?

**Comment by robinlioret** [comment]
I'm facing the same issue. With the same configuration as you. But i'm a newbie regarding javascript. Did you find a solution ?

* I'm having an error with the ` import mermaid from "mermaid"`  
  `Uncaught TypeError: The specifier “mermaid” was a bare specifier, but was not remapped to anything.`  
  So i switched a complete cdn url: https://cdn.jsdelivr.net/npm/mermaid@11.4.1/+esm

* I wonder if the pack reaches mermaid. Is it the "same" mermaid in this module than the one loaded by mkdocs ?

**Comment by robinlioret** [comment]
I found a way to make it work: https://github.com/fralau/mkdocs-mermaid2-plugin/issues/111#issuecomment-2473992133

(referring to the mermaid2 plugin documentation : https://mkdocs-mermaid2.readthedocs.io/en/latest/library/#using-extra_javascript)


To make it work, we need to install the [mermaid2 plugin](https://mkdocs-mermaid2.readthedocs.io/en/latest/#installation)

Here is the set up:

```javascript
# docs/_javascript/register-mermaid-icons.js
import mermaid from 'https://unpkg.com/mermaid@11.4.0/dist/mermaid.esm.min.mjs'
document$.subscribe(function() {
  console.log("Initialize third-party libraries here")
  mermaid.registerIconPacks([
    {
      name: 'logos',
      loader: () =>
        fetch('https://unpkg.com/@iconify-json/logos@1/icons.json').then((res) => res.json()),
    }
  ])
  mermaid.run({querySelector: '.diagram'});
});
```

```yaml
# mkdocs.yaml
site_name: My Docs
theme:
  name: material

extra_javascript:
  - path: _javascript/register-mermaid-icons.js
    type: module

plugins:
  - search
  - mermaid2:
      version: 11.4.0
      arguments:
        securityLevel: 'loose'
        startOnLoad: false

markdown_extensions:
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: diagram
          format: !!python/name:mermaid2.fence_mermaid
``` 

I got the following result:

docs/index.md
````markdown
# Welcome to MkDocs

```mermaid
architecture-beta
    group api(logos:aws-lambda)[API]

    service db(logos:aws-aurora)[Database] in api
    service disk1(logos:aws-glacier)[Storage] in api
    service disk2(logos:aws-s3)[Storage] in api
    service server(logos:aws-ec2)[Server] in api

    db:L -- R:server
    disk1:T -- B:server
    disk2:T -- B:db
```
````
![image](https://github.com/user-attachments/assets/ab4d3a96-794b-40ed-bd53-c2c429b295fd)

Let me know if it worked for you.

**Comment by segfly** [comment]
I ran into the same issue but took a different approach with my solution because, in my opinion, the previous suggestions had the following drawbacks:

- If a different class besides `mermaid` is used, then the Material for MkDocs styling is not available for the mermaid diagrams. This means losing theme color and light/dark mode support without a lot of extra work.
- There are performance/rendering considerations (like rendering in a shadow DOM) that Material for MKDocs incorporated from the `pymdown-extensions`. I don't want to re-implement these.
- I'd rather not bring in another plugin and extra configuration if unnecessary.

For me, the best solution is to use the built-in Mermaid support for Material for MkDocs, but somehow ensure that the icons are loaded before the [`mermaid.initialize()`](https://github.com/squidfunk/mkdocs-material/blob/9e6c0674d65a5ad5322c673707d1fda62c1a5e9d/src/templates/assets/javascripts/components/content/mermaid/index.ts#L93) call is made by Material for MkDocs.

However, I don't want to have to fork/build my own version of Material for MkDocs.

So what to do? I decided to hook into the `mermaid` object lifecycle by creating a property setter to intercept creation of the global `mermaid` object when it is loaded. This allowed me to call the `mermaid.registerIconPacks` before any of the other mermaid initialization or rendering code was called.

I think this solution is more elegant, preserves full Material for MkDocs support, seems to work in any browser I tested, and allows for additional Mermaid config changes prior to initialization if needed.

Below is the code. No other plugins are necessary:

**mkdocs.yaml**
```yaml
extra_javascript:
  - path: assets/javascript/extra.js
markdown_extensions:
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:pymdownx.superfences.fence_code_format
```

**extra.js**
```js
(function () {
    // Begin async loading icon packs ASAP so they don't slow down the diagram rendering 
    const iconify_logos = fetch('https://unpkg.com/@iconify-json/logos@1/icons.json').then((res) => res.json())
        .catch(e => console.error("Failed to fetch Mermaid icon pack:", e));

    // Example of loading additional packs if needed
    const iconify_material = fetch('https://unpkg.com/@iconify-json/material-symbols@1/icons.json').then((res) => res.json())
        .catch(e => console.error("Failed to fetch Mermaid icon pack:", e));

    // Intercept Mermaid loading into the global scope so we can register
    // the icon packs before Mermaid is initialized by Material for MkDocs
    Object.defineProperty(window, 'mermaid', {
        configurable: true,
        set: async function (value) {
            delete window.mermaid;
            window.mermaid = value;

            try {
                value.registerIconPacks([
                    { name: 'logos', loader: () => iconify_logos },
                    { name: 'material-symbols', loader: () => iconify_material },
                ]);
            } catch (e) {
                console.error("Error registering Mermaid icon pack:", e);
            }
        }
    });
})();
```

**example.md**
````markdown
``` mermaid
architecture-beta
    group aws(logos:aws-lambda)[Lambda]
    service support(material-symbols:medical-services)[Support] in aws
```
````

**Dark Mode**
![image](https://github.com/user-attachments/assets/06e7e2b7-bf81-4b04-9605-abed4a2b82ca)

**Light Mode**
![image](https://github.com/user-attachments/assets/5ec8f9d6-0c01-478e-8395-5effb420e3d4)


Hope this helps!

**Reply by serpro69** [reply]
Thanks a lot for this! This is the best solution I've found so far that works and does not require additional plugins

**Reply by onemargaro** [reply]
Thanks its really what i looking for and was easy to do it, just a small step i tried to put the assets folder in root folder and needs to be inside docs like
<img width="383" height="110" alt="image" src="https://github.com/user-attachments/assets/1cb67f0e-8c9b-4041-8d50-034e00b17a2d" />
